core/src/test/java/org/infinispan/partitionhandling/TransactionalPartitionAndMergeTest.java

```diff
@@ -0,0 +1,794 @@
+
+      mergeCluster();
+      finalAsserts(PESSIMISTIC_TX_CACHE_NAME, keyInfo, commit ? FINAL_VALUE : INITIAL_VALUE);
+   }
+
+   private void doTestSplitBeforeWithOptimistic(boolean commit) throws Exception {
+      //the transaction is successfully prepare and then the split happens before the commit phase starts.
+      waitForClusterToForm(OPTIMISTIC_TX_CACHE_NAME);
+      final KeyInfo keyInfo = createKeys(OPTIMISTIC_TX_CACHE_NAME);
+      final Cache<Object, String> originator = cache(0, OPTIMISTIC_TX_CACHE_NAME);
+
+      final DummyTransactionManager transactionManager = (DummyTransactionManager) originator.getAdvancedCache().getTransactionManager();
+      transactionManager.begin();
+      final DummyTransaction transaction = transactionManager.getTransaction();
+      keyInfo.putFinalValue(originator);
+      AssertJUnit.assertTrue(transaction.runPrepare());
+      transactionManager.suspend();
+
+      SplitMode.BOTH_DEGRADED.split(this);
+
+      transactionManager.resume(transaction);
+      transaction.runCommit(!commit);
+
+      if (!commit) {
+         assertEventuallyNotLocked(cache(1, OPTIMISTIC_TX_CACHE_NAME), keyInfo.getKey1());
+      } else {
+         assertLocked(cache(1, OPTIMISTIC_TX_CACHE_NAME), keyInfo.getKey1());
+      }
+      assertLocked(cache(2, OPTIMISTIC_TX_CACHE_NAME), keyInfo.getKey2());
+
+      mergeCluster();
+      finalAsserts(OPTIMISTIC_TX_CACHE_NAME, keyInfo, commit ? FINAL_VALUE : INITIAL_VALUE);
+   }
+
+   private void doOptimisticTxTest(final SplitMode splitMode, final TransactionPhase phase, boolean txFail, boolean discard) throws Exception {
+      waitForClusterToForm(OPTIMISTIC_TX_CACHE_NAME);
+
+      final KeyInfo keyInfo = createKeys(OPTIMISTIC_TX_CACHE_NAME);
+      final Cache<Object, String> originator = cache(0, OPTIMISTIC_TX_CACHE_NAME);
+      final FilterCollection filterCollection = createFilters(OPTIMISTIC_TX_CACHE_NAME, discard, phase.getCommandClass(), splitMode);
+
+      Future<Void> put = fork(new Callable<Void>() {
+         @Override
+         public Void call() throws Exception {
+            final DummyTransactionManager transactionManager = (DummyTransactionManager) originator.getAdvancedCache().getTransactionManager();
+            transactionManager.begin();
+            keyInfo.putFinalValue(originator);
+            final DummyTransaction transaction = transactionManager.getTransaction();
+            transaction.runPrepare();
+            transaction.runCommit(phase == TransactionPhase.ROLLBACK);
+            transaction.throwRollbackExceptionIfAny();
+            return null;
+         }
+      });
+
+      filterCollection.await(30, TimeUnit.SECONDS);
+      splitMode.split(this);
+      filterCollection.unblock();
+
+      try {
+         put.get();
+         assertFalse(txFail);
+      } catch (ExecutionException e) {
+         assertTrue(txFail);
+      }
+
+      switch (phase) {
+         case PREPARE:
+            switch (splitMode) {
```

Comment:
I'm not a big fan of nested switch statements, how about extracting separate methods from the top-level cases, like `assertPrepareLocks` and `assertRollbackLocks`?